ops/internal/deployer/deployer.go

```diff
@@ -0,0 +1,161 @@
+	var versions map[string]string
+	if err := json.Unmarshal(versionsData, &versions); err != nil {
+		return "", fmt.Errorf("failed to parse versions.json: %w", err)
+	}
+
+	deployerVersion, ok := versions[contractsKey]
+	if !ok {
+		return "", fmt.Errorf("no deployer version found for contracts: %s", contractsKey)
+	}
+	d.lgr.Info("Found deployer version", "version", deployerVersion)
+	d.deployerVersion = deployerVersion
+
+	// Run the build-deployer script with the appropriate deployer version
+	d.lgr.Info("Running build-deployer script")
+	scriptPath := filepath.Join(d.rootDir, "scripts", "build-deployer.sh")
+	cmd := exec.Command(scriptPath, deployerVersion)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return "", fmt.Errorf("failed to run build-deployer script: %w", err)
+	}
+
+	return deployerVersion, nil
+}
+
+// SetupStateAndIntent prepares the deployer environment by creating merged state and intent files
+// in the specified working directory.
+func (d *OpDeployer) SetupStateAndIntent(workdir string) error {
+	// Read the state file
+	state, err := ReadOpaqueMappingFile(d.inputStatePath)
+	if err != nil {
+		return fmt.Errorf("failed to read state file: %w", err)
+	}
+
+	// Determine version based on deployer version and call appropriate merge function
+	var mergedIntent, mergedState OpaqueMapping
+	// Parse deployer version to determine which merge function to use
+	if strings.Contains(d.deployerVersion, ".3.") {
+		mergedIntent, mergedState, err = MergeStateV3(state)
+	} else {
+		mergedIntent, mergedState, err = MergeStateV2(state)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to merge state: %w", err)
+	}
+
+	// Write state.json in the temp directory
+	stateJSON, err := json.MarshalIndent(mergedState, "", "  ")
+	if err != nil {
+		return fmt.Errorf("failed to marshal state to JSON: %w", err)
+	}
+	stateTempPath := filepath.Join(workdir, "state.json")
+	if err := os.WriteFile(stateTempPath, stateJSON, 0o644); err != nil {
+		return fmt.Errorf("failed to write state to temp file: %w", err)
+	}
+	d.lgr.Info("Wrote state to temporary file", "path", stateTempPath)
+
+	// Write intent.toml in the temp directory
+	intentTOML, err := toml.Marshal(mergedIntent)
+	if err != nil {
+		return fmt.Errorf("failed to marshal intent to TOML: %w", err)
+	}
+	intentTempPath := filepath.Join(workdir, "intent.toml")
+	if err := os.WriteFile(intentTempPath, intentTOML, 0o644); err != nil {
+		return fmt.Errorf("failed to write intent to temp file: %w", err)
+	}
+	d.lgr.Info("Wrote intent to temporary file", "path", intentTempPath)
+
+	return nil
+}
+
+// GenerateGenesis runs op-deployer binary to generate a genesis
+func (d *OpDeployer) GenerateGenesis(workdir string) (*core.Genesis, error) {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get home directory: %w", err)
+	}
+	deployerPath := filepath.Join(homeDir, ".cache/binaries", d.deployerVersion, "op-deployer")
```

Comment:
These few lines could be factored into a method `OpDeployer.getExePath()" so we can reuse them in other methods.